### PR TITLE
E2E test for Elastic Agent retrying downloads

### PR DESCRIPTION
**Pull request by ycombinator**
## What does this PR do?

This PR adds an E2E test to ensure that Elastic Agent retries downloads while upgrading.

## Why is it important?

To ensure that upgrades work as expected with transient download failures.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~I have made corresponding changes to the documentation~
- [ ] ~I have made corresponding change to the default configuration files~
- [ ] ~I have added tests that prove my fix is effective or that my feature works~
- [ ] ~I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~
- [x] I have added an integration test or an E2E test

### CODE REVIEW

File: pkg/testing/fixture.go [hunk 1]
```diff
@@ -374,7 +374,7 @@ func (f *Fixture) getFetcherCache() *fetcherCache {
 
 func (f *Fixture) prepareComponents(workDir string, components ...UsableComponent) error {
 	if len(components) == 0 {
-		f.t.Logf("Components where not modified from the fetched artifact")
+		f.t.Logf("Components were not modified from the fetched artifact")
```
Comment (ycombinator) [inline pkg/testing/fixture.go:377]: This change is unrelated to the E2E test introduced in this PR; it's just a typo I noticed while running the E2E test in this PR.

**Review by blakerouse — APPROVED** [review]
Looks good, just a comment on `isolate`. Not a blocker.

File: testing/integration/upgrade_retry_download_test.go [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+	"github.com/elastic/elastic-agent/internal/pkg/release"
+	atesting "github.com/elastic/elastic-agent/pkg/testing"
+	"github.com/elastic/elastic-agent/pkg/testing/define"
+	"github.com/elastic/elastic-agent/pkg/testing/tools"
+)
+
+func TestElasticAgentUpgradeRetryDownload(t *testing.T) {
+	info := define.Require(t, define.Requirements{
+		Local:   false, // requires Agent installation
+		Isolate: true,  // requires Agent installation and modifying /etc/hosts
```
[the remark this answers is not in the document]
Comment (ycombinator) [reply testing/integration/upgrade_retry_download_test.go:34]: Yeah, good question.  Will non-isolated tests run in parallel on the same host?  If that's a possibility, then this test should be isolated as the jacked up `/etc/hosts` might cause other tests to fail.  But if non-isolated tests are guaranteed to run serially on the same host, then we don't need this test to be isolated.

[the remark this answers is not in the document]
Comment (blakerouse) [reply testing/integration/upgrade_retry_download_test.go:34]: The tests are executed through `go test`. I believe unless you call like `t.Parallel` that it will be run serially.

**Review by blakerouse — APPROVED** [review]
Looks good!
